tensorflow/core/kernels/segment_reduction_ops_impl.h

```diff
@@ -957,6 +957,125 @@ class SparseSegmentSqrtNGradOp
                                                      true /*is_sqrtn*/) {}
 };
 
+#if GOOGLE_CUDA || TENSORFLOW_USE_ROCM
+template <class Device, class T, class Tindex, bool has_num_segments>
+class SparseSegmentSumGpuOp : public AsyncOpKernel {
+ public:
+  explicit SparseSegmentSumGpuOp(OpKernelConstruction* context)
+      : AsyncOpKernel(context){};
+
+  void ComputeAsync(OpKernelContext* context, DoneCallback done) override {
+    const Tensor& input_data = context->input(0);
+    const Tensor& indices = context->input(1);
+    const Tensor& segment_ids = context->input(2);
+
+    OP_REQUIRES_ASYNC(context, TensorShapeUtils::IsVector(indices.shape()),
+                      errors::InvalidArgument("indices should be a vector."),
+                      done);
+    const int64 num_indices = indices.NumElements();
+
+    OP_REQUIRES_ASYNC(
+        context, TensorShapeUtils::IsVector(segment_ids.shape()),
+        errors::InvalidArgument("segment_ids should be a vector."), done);
+
+    OP_REQUIRES_ASYNC(
+        context, num_indices == segment_ids.NumElements(),
+        errors::InvalidArgument("indices and segment_ids should have"
+                                "same length."),
+        done);
+
+    ScratchSpace<Index> output_rows_host(context, 1, /* on_host */ true);
+    auto stream = context->op_device_context()->stream();
+
+    if (has_num_segments) {
+      const Tensor& num_segments = context->input(3);
+      se::DeviceMemoryBase num_segments_device(
+          const_cast<Tensor&>(num_segments).template flat<Index>().data()
+      );
+      OP_REQUIRES_ASYNC(
+          context,
+          stream
+              ->ThenMemcpy(output_rows_host.mutable_data(), num_segments_device, sizeof(Index)).ok(),
+          errors::Internal(
+              "SparseSegmentSumGpuOp: failed to copy num_segments to host."),
+          done);
```

Comment:
If I'm reading the code correctly, the correct way to do this is to use the [`HostMemory` annotation](https://github.com/tensorflow/tensorflow/blob/00abe72aee36b7661ef0746d6349fdc7389ca705/tensorflow/core/framework/op_kernel.h#L1354).  The runtime will then copy that input to the host before running the `OpKernel`.

Reply:
It seems more concise!
[registration](https://github.com/tensorflow/tensorflow/pull/45085/commits/a119214973391761cda0c74698ed239456361b98#diff-157a31ffb67895d72ee7825a151b5aecde67d8450ed67952e4cddd50c65ae3bcR140)
[simpler impl](https://github.com/tensorflow/tensorflow/pull/45085/commits/a119214973391761cda0c74698ed239456361b98#diff-d29812caca45335066ddacf8fb06afc416f1f557bc446068ebaa163495e0b03bL992)